File: src/host/lib/fragment.ml
(**********************************************************************)
(*                                                                    *)
(*              This file is part of the RFSM package                 *)
(*                                                                    *)
(*  Copyright (c) 2018-present, Jocelyn SEROT.  All rights reserved.  *)
(*                                                                    *)
(*  This source code is licensed under the license found in the       *)
(*  LICENSE file in the root directory of this source tree.           *)
(*                                                                    *)
(**********************************************************************)

type t = {
  inps: (string * string) list; (* id, type expr *)
  outps: (string * string) list; (* id, type expr *)
  vars: (string * string) list; (* id, type expr *)
  obj: string; (* fragment to analyse; ex ["guard x=1"]  *)
} [@@deriving show]

let of_json (json : Yojson.Basic.t) : t =
  let open Yojson.Basic.Util in
  { inps = json |> member "inps" |> to_list
           |> List.map (fun j -> (j |> member "id" |> to_string,
                                  j |> member "type" |> to_string));
    outps = json |> member "outps" |> to_list
           |> List.map (fun j -> (j |> member "id" |> to_string,
                                  j |> member "type" |> to_string));
    vars = json |> member "vars" |> to_list
           |> List.map (fun j -> (j |> member "id" |> to_string,
                                  j |> member "type" |> to_string));
    obj = json |> member "obj" |> to_string; }

(* let from_json json = 
 *   let open Yojson.Basic.Util in
 *   let extract_iov j = member "id" j |> to_string, member "type" j |> to_string in
 *   let extract_iovs name j = j |> member name |> to_list |> List.map extract_iov in
 *   { inps = extract_iovs "inps" json;
 *     outps = extract_iovs "outps" json;
 *     vars =  extract_iovs "vars" json;
 *     obj = json |> member "obj" |> to_string; } *)

let of_string s =
  of_json (Yojson.Basic.from_string s)

let to_json (f : t) : Yojson.Basic.t =
  let pair_list_to_yojson lst =
  `List (
    List.map (fun (a, b) ->
      `Assoc [("id", `String a); ("type", `String b)]
      ) lst
    ) in
  `Assoc [
    ("inps", pair_list_to_yojson f.inps);
    ("outps", pair_list_to_yojson f.outps);
    ("vars", pair_list_to_yojson f.vars);
    ("obj", `String f.obj)
  ]

let to_string f =
  Yojson.Basic.pretty_to_string (to_json f)

File: src/guests/std/tests/server/client/fragment.cpp
/***********************************************************************/
/*                                                                     */
/*       This file is part of the Grasp software package               */
/*                                                                     */
/*  Copyright (c) 2019-present, Jocelyn SEROT (jocelyn.serot@uca.fr)   */
/*                       All rights reserved.                          */
/*                                                                     */
/*    This source code is licensed under the license found in the      */
/*      LICENSE file in the root directory of this source tree.        */
/*                                                                     */
/***********************************************************************/

#include "fragment.h"

#include <QString>
#include <QPair>
#include <QList>
#include <QJsonObject>
#include <QJsonArray>

QJsonObject Fragment::toJson() const
{
        QJsonObject json;
        json["inps"] = listToJsonArray(inps);
        json["outps"] = listToJsonArray(outps);
        json["vars"] = listToJsonArray(vars);
        json["obj"] = obj;
        return json;
}

Fragment Fragment::fromJson(const QJsonObject &json)
{
        return Fragment(
            jsonArrayToList(json["inps"].toArray()),
            jsonArrayToList(json["outps"].toArray()),
            jsonArrayToList(json["vars"].toArray()),
            json["obj"].toString()
        );
}

QJsonArray Fragment::listToJsonArray(const QList<QPair<QString, QString>> &list)
{
        QJsonArray arr;
        for (const auto &p : list) {
            QJsonObject o;
            o["id"] = p.first;
            o["type"] = p.second;
            arr.append(o);
        }
        return arr;
}

QList<QPair<QString, QString>> Fragment::jsonArrayToList(const QJsonArray &arr)
{
        QList<QPair<QString, QString>> list;
        for (const auto &val : arr) {
            QJsonObject o = val.toObject();
            list.append(qMakePair(o["id"].toString(), o["type"].toString()));
        }
        return list;
}
